File: python/Leetcode-Roadmap/04_Binary_Search/koko_eating_bananas.py
import math

class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        # set left & right pointers
        l, r = 1, max(piles)

        # initialise result to max number in list
        res = r

        while l <= r:
            # get the value of k (middle value in list)
            k = (l + r) // 2

            totalTime = 0
            # loop through each item in list
            for p in piles:
                # divide list item by k
                totalTime += math.ceil(float(p) / k)

            # update result if hours less or equal to input
            if totalTime <= h:
                res = k
                # update right pointer to find left portion of list
                r = k - 1
            else:
                # update left pointer to find right side of list
                l = k + 1
        
        # return the result
        return res
    
# implementation
answer = Solution()

print("Leetcode 875: Koko Eating Bananas\n")

# Example 1
piles_1 = [3, 6, 7, 11]
hours_1 = 8
print(f"Example 1: piles = {piles_1}, h = {hours_1}")
print("Output:", answer.minEatingSpeed(piles_1, hours_1))

# Example 2
piles_2 = [30, 11, 23, 4, 20]
hours_2 = 5
print(f"Example 2: piles = {piles_2}, h = {hours_2}")
print("Output:", answer.minEatingSpeed(piles_2, hours_2))

# Example 3
hours_3 = 6
print(f"Example 2: piles = {piles_2}, h = {hours_3}")
print("Output:", answer.minEatingSpeed(piles_2, hours_3))

File: python/Leetcode-Roadmap/04_Binary_Search/kokoEatingBananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    // set left & right pointers
    let left = 1, right = Math.max(...piles);
    // set result to highest number (this is only temporary)
    let result = right;
    
    while (left <= right) {
        // get the value of k 
        const k = Math.floor((left + right) / 2);
        let totalTime = 0;
        
        // loop through each item in list
        for (const pile of piles) {
            // update total time by dividing pile by k
            // use Math.ceil() to round up to nearest number
            totalTime += Math.ceil(pile / k);
        }
        
        // binary search
        
        // update the result if totalTime is less or equal to hours
        if (totalTime <= h) {
            result = k;
            // update right pointer to search the left side of list
            right = k - 1;
        } else {
            // update left pointer to search the right side of list
            left = k + 1;
        }
        
    }
    // return the result after the loop has finished
    return result;
    
};

// implementation
console.log("Leetcode 875: Koko Eating Bananas\n");

// Example 1
const piles1 = [3, 6, 7, 11];
var hours1 = 8;
console.log(`Example 1: piles = [${piles1}], h = ${hours1}`);
console.log("Output:", minEatingSpeed(piles1, hours1));

// Example 2
const piles2 = [30, 11, 23, 4, 20];
var hours2 = 5;
console.log(`Example 2: piles = [${piles2}], h = ${hours2}`);
console.log("Output:", minEatingSpeed(piles2, hours2));

// Example 3
var hours3 = 6;
console.log(`Example 3: piles = [${piles2}], h = ${hours3}`);
console.log("Output:", minEatingSpeed(piles2, hours3));